### PR TITLE
#666 Improve IdentityComparer logic.

### PR DESCRIPTION
**Pull request by mafurman**
Implements #666.

### CODE REVIEW

File: test/Microsoft.IdentityModel.Tests/IdentityComparer.cs [hunk 1]
```diff
@@ -37,108 +37,83 @@
...
 {
-    public class CompareContext
-    {       
-        List<string> _diffs = new List<string>();
+    public class IdentityComparer
+    {
+        private static readonly Dictionary<string, Func<Object, object, CompareContext, bool>> _equalityDict =
+            new Dictionary<string, Func<Object, object, CompareContext, bool>>()
+        {
+            {typeof(IEnumerable<Claim>).ToString(), AreClaimsEnumsEqual },
```
Comment (brentschmaltz) [inline test/Microsoft.IdentityModel.Tests/IdentityComparer.cs]: nit: add space after '{'. {typeof -> { typeof

File: test/Microsoft.IdentityModel.Tests/IdentityComparer.cs [hunk 2]
```diff
@@ -37,108 +37,83 @@
...
+            {typeof(IEnumerable<ClaimsIdentity>).ToString(), AreClaimsIdentitiesEnumsEqual },
+            {typeof(IEnumerable<string>).ToString(), AreStringEnumsEqual },
+            {typeof(IEnumerable<SecurityKey>).ToString(), AreSecurityKeyEnumsEqual },
+            {typeof(Newtonsoft.Json.Linq.JArray).ToString(), AreJArraysEqual },
```
Comment (brentschmaltz) [inline test/Microsoft.IdentityModel.Tests/IdentityComparer.cs]: Check if .ToString() is needed. Every object has that method.
If not needed, just remove it.

File: test/Microsoft.IdentityModel.Tests/IdentityComparer.cs [hunk 3]
```diff
@@ -37,108 +37,83 @@
...
+            {typeof(IDictionary<string, string>).ToString(), AreStringDictionariesEqual},
+            {typeof(Dictionary<string, object>).ToString(), AreStringObjectDictionariesEqual },
+            {typeof(Dictionary<string, object>.ValueCollection).ToString(), AreValueCollectionsEqual },
+            {typeof(IEnumerable<object>).ToString(), AreObjectEnumsEqual },
+            {typeof(Collection<SecurityKey>).ToString(), ContinueCheckingEquality },
+        };
+
+        // Checks if either object is null or if both are references of each other to see if we can terminate 
+        // comparison early. Returns null if neither case applies.
+        private static bool? AreObjectsNullOrReferences(Object o1, Object o2, CompareContext context)
```
Comment (brentschmaltz) [inline test/Microsoft.IdentityModel.Tests/IdentityComparer.cs]: please change parameter names to object1, object2.
using English type names helps in readability.

(please change the other instances)

File: test/Microsoft.IdentityModel.Tests/IdentityComparer.cs [hunk 4]
```diff
@@ -37,108 +37,83 @@
...
+        private static readonly Dictionary<string, Func<Object, object, CompareContext, bool>> _equalityDict =
+            new Dictionary<string, Func<Object, object, CompareContext, bool>>()
+        {
+            {typeof(IEnumerable<Claim>).ToString(), AreClaimsEnumsEqual },
+            {typeof(IEnumerable<ClaimsIdentity>).ToString(), AreClaimsIdentitiesEnumsEqual },
+            {typeof(IEnumerable<string>).ToString(), AreStringEnumsEqual },
+            {typeof(IEnumerable<SecurityKey>).ToString(), AreSecurityKeyEnumsEqual },
+            {typeof(Newtonsoft.Json.Linq.JArray).ToString(), AreJArraysEqual },
+            {typeof(string).ToString(), AreStringsEqual },
+            {typeof(IDictionary<string, string>).ToString(), AreStringDictionariesEqual},
```
Comment (brentschmaltz) [inline test/Microsoft.IdentityModel.Tests/IdentityComparer.cs]: slight inconsistent naming here.
Either 'AreStringStringDictionariesEqual' OR 'AreObjectDictionariesEqual'.

File: test/Microsoft.IdentityModel.Tests/IdentityComparer.cs [hunk 5]
```diff
@@ -37,108 +37,83 @@
...
-
-    public class IdentityComparer
-    {
         public static bool AreEnumsEqual<T>(IEnumerable<T> t1, IEnumerable<T> t2, CompareContext context, Func<T, T, CompareContext, bool> areEqual)
         {
-            if (t1 == null && t2 == null)
-                return true;
+            // Check if either t1 or t2 are null or references of each other to see if we can terminate early.
+            // Is this necessary if we're already checking this in AreEqual()?
+            bool? res = AreObjectsNullOrReferences(t1,t2,context);
```
Comment (brentschmaltz) [inline test/Microsoft.IdentityModel.Tests/IdentityComparer.cs]: naming: use 'result' instead or 'res'.

(please change the other instances)

File: test/Microsoft.IdentityModel.Tests/IdentityComparer.cs [hunk 6]
```diff
@@ -37,108 +37,83 @@
...
-
-            if (t2 == null)
-            {
-                context.Diffs.Add($"t1 != null, t2 == null, (type: {t1.GetType()}), value: {t1.ToString()}).");
-                return false;
-            }
-
-            if (ReferenceEquals(t1, t2))
-                return true;
+            if (res != null)
```
Comment (brentschmaltz) [inline test/Microsoft.IdentityModel.Tests/IdentityComparer.cs]: for nullablies use a different pattern:
if (res.HasValue)
    return res.Value;

(change the other instances)

File: test/Microsoft.IdentityModel.Tests/IdentityComparer.cs [hunk 7]
```diff
@@ -392,234 +363,41 @@ public static bool AreEqual(object t1, object t2)
...
+
+            string inter;
+            // Use a special function for comparison if required by the specific class of the object.
+            if (_equalityDict.TryGetValue(t1.GetType().ToString(), out Func<Object, object, CompareContext, bool> areEqual))
             {
-                context.Diffs.Add($"t1 != null, t2 == null, (type: {t1.GetType()}), value: {t1.ToString()}).");
-                return false;
+                return areEqual(t1, t2, localContext);
+            // Check if any of the interfaces that the class uses require a special function.
+            } else if ((inter = t1.GetType().GetInterfaces().Select(t => t.ToString()).Intersect(_equalityDict.Keys).FirstOrDefault()) != null)
```
Comment (brentschmaltz) [inline test/Microsoft.IdentityModel.Tests/IdentityComparer.cs]: put the 'else if' on separate line.

File: test/Microsoft.IdentityModel.Tests/IdentityComparer.cs (hunk 2 above)
Comment (mafurman) [inline test/Microsoft.IdentityModel.Tests/IdentityComparer.cs]: What do you mean? I don't think it's possible for ToString() to be called implicitly here?

File: test/Microsoft.IdentityModel.Tests/IdentityComparer.cs (hunk 4 above)
Comment (mafurman) [inline test/Microsoft.IdentityModel.Tests/IdentityComparer.cs]: Just changed it to 'AreObjectDictionariesEqual.' I feel like it's just as readable while being less verbose.

File: test/Microsoft.IdentityModel.Tests/IdentityComparer.cs (hunk 1 above)
Comment (mafurman) [inline test/Microsoft.IdentityModel.Tests/IdentityComparer.cs]: Done!

File: test/Microsoft.IdentityModel.Tests/IdentityComparer.cs (hunk 3 above)
Comment (mafurman) [inline test/Microsoft.IdentityModel.Tests/IdentityComparer.cs]: Done!

File: test/Microsoft.IdentityModel.Tests/IdentityComparer.cs (hunk 5 above)
Comment (mafurman) [inline test/Microsoft.IdentityModel.Tests/IdentityComparer.cs]: Done.

File: test/Microsoft.IdentityModel.Tests/IdentityComparer.cs (hunk 7 above)
Comment (mafurman) [inline test/Microsoft.IdentityModel.Tests/IdentityComparer.cs]: Done!

File: test/Microsoft.IdentityModel.Tests/IdentityComparer.cs (hunk 6 above)
Comment (mafurman) [inline test/Microsoft.IdentityModel.Tests/IdentityComparer.cs]: Done!

**Comment by brentschmaltz** [comment]
:shipit: rebase, squash and commit.

**Comment by mafurman** [comment]
Pushed directly to dev.

**Pull request closed without merging**